peer.go

```diff
@@ -274,201 +303,242 @@ func (p *Peer) String() string {
 
-	operation := func() (wire.Message, error) {
+	go func() {
 		msg, _, err := wire.ReadMessage(r, pver, bsvnet)
-		if err != nil {
-			return nil, err
-		}
-		return msg, nil
-	}
+		readMessageFinished <- readMessageResult{msg, err}
+	}()
 
-	notify := func(err error, nextTry time.Duration) {
-		if errors.Is(err, io.EOF) {
-			p.logger.Error("Failed to read message: EOF", slog.String("next try", nextTry.String()), slog.String(errKey, err.Error()))
-		} else {
-			p.logger.Error("Failed to read message", slog.String("next try", nextTry.String()), slog.String(errKey, err.Error()))
+	// ensure read message doesn't block
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case readMsg := <-readMessageFinished:
+		if readMsg.err != nil {
+			return nil, readMsg.err
 		}
+		return readMsg.msg, nil
 	}
+}
 
-	msg, err := backoff.RetryNotifyWithData(operation, policyContext, notify)
-	if err != nil {
-		return nil, err
+func (p *Peer) readRetry(ctx context.Context, r io.Reader, pver uint32, bsvnet wire.BitcoinNet) (wire.Message, error) {
+	msg, err := p.readMessage(ctx, r, pver, bsvnet)
+	if err == nil {
+		return msg, nil
 	}
 
-	return msg, nil
-}
+	if errors.Is(err, io.EOF) {
+		p.logger.Error("Failed to read message: EOF", slog.String(errKey, err.Error()))
+	} else {
+		p.logger.Error("Failed to read message", slog.String(errKey, err.Error()))
+	}
 
-func (p *Peer) startReadHandler() {
-	ctx, cancel := context.WithCancel(context.Background())
-	p.cancelReadHandler = cancel
+	counter := 0
+	ticker := time.NewTicker(p.retryReadWriteMessageInterval)
 
-	p.logger.Info("Starting read handler")
+	var readMessageErr error
+	for {
+		counter++
+		if counter >= retryReadWriteMessageAttempts {
+			break
+		}
 
-	go func(cancelCtx context.Context) {
-		defer func() {
-			p.logger.Info("Shutting down read handler")
-		}()
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-ticker.C:
+			msg, readMessageErr = p.readMessage(ctx, r, pver, bsvnet)
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					p.logger.Error("Failed to read message: EOF", slog.String(errKey, err.Error()))
+					continue
+				}
 
-		readConn := p.readConn
-		var msg wire.Message
-		var err error
+				p.logger.Error("Failed to read message", slog.String(errKey, err.Error()))
+				continue
+			}
 
-		if readConn == nil {
-			p.logger.Error("no connection")
-			return
+			return msg, nil
 		}
+	}
 
-		reader := bufio.NewReader(&io.LimitedReader{R: readConn, N: p.maximumMessageSize})
-		for {
-			select {
-			case <-cancelCtx.Done():
-				return
-			default:
-				msg, err = p.readRetry(cancelCtx, reader, wire.ProtocolVersion, p.network)
-				if err != nil {
-					if errors.Is(err, context.Canceled) {
-						p.logger.Info("Retrying to read cancelled")
-						return
-					}
+	return nil, readMessageErr
+}
 
-					p.logger.Error("Retrying to read failed", slog.String(errKey, err.Error()))
+func (p *Peer) startReadHandler(cancelCtx context.Context) {
```

Comment:
Shouldn't this just be called `ctx` instead of `cancelCtx`? It's just context and not a cancel function.